### PR TITLE
initial commit of quadratic voting code

### LINKED ISSUE #104: Create a quadratic voting smart contract
- [x] Create a new git branch for this ticket and switch to it
- [x] Create an `ethereum` directory in the repository's root
- [x] Create a `voting` sub-directory within the `ethereum` directory
- [x] Check in a draft quadratic voting smart contract into the `voting` directory
- [x] Create a pull request and ask me, @anonymoussprocket and @bhaskarkishore for reviews
- [ ] Incorporate all feedback and finalize your contract for approval and merging

### PR DESCRIPTION
**Pull request by dwariyar**
Closes #104.

### CODE REVIEW

File: Etherium/Voting/_Quadratic_Voting.sol
Comment (vishakh) [inline Etherium/Voting/_Quadratic_Voting.sol:1]: Using either the [Remix IDE](https://remix.ethereum.org/#lang=en&optimize=false&runs=200&evmVersion=null) or [ethers.js](https://docs.ethers.org/v5/), could you please create unit tests to establish that the contract behaves as expected?

File: Etherium/Voting/_Quadratic_Voting.sol [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            voters[voter].voterCreditSpent != 100,
+            "The voter already spent their voting credits."
+        );
+        require(voters[voter].voterCreditBank == 0);
+        voters[voter].voterCreditBank = 100;
+        voters[voter].voterCreditSpent = 0;
+    }
+
+    //Takes square root of input value
+    function sqrt(uint y) internal pure returns (uint z) {
```
Comment (vishakh) [inline Etherium/Voting/_Quadratic_Voting.sol:54]: Try to use a standard library such as [SafeMath](https://docs.openzeppelin.com/contracts/2.x/api/math) to ensure that the math is robust. This is preferable to rolling your own utility functions.

File: Etherium/Voting/_Quadratic_Voting.sol [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        winnerName_ = proposals[chooseWinningProposal()].name;
+    }
+
+}
+
+//converts from string to bytes32 format
+contract converter {
+
+    function stringToBytes32(string memory source) public pure returns (bytes32 result) {
```
Comment (vishakh) [inline Etherium/Voting/_Quadratic_Voting.sol:107]: Try to use a library function here.

**Review by vishakh — CHANGES_REQUESTED** [review]

File: Etherium/Voting/_Quadratic_Voting.sol
Comment (vishakh) [inline Etherium/Voting/_Quadratic_Voting.sol:1]: "Ethereum" is misspelled in the path.

File: Ethereum/Voting/_Quadratic_Voting.sol [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+
+function getVoterCreditSpent(address voter) public view returns (uint256) {
+    return voters[voter].voterCreditSpent;
+}
+
+
+function getVoteCount(bytes32 proposal) public view returns (uint256) {
+    //return proposals[proposalNames[j]].voteCount;
```
Comment (vishakh) [inline Ethereum/Voting/_Quadratic_Voting.sol:111]: Could we either implement this function or remove it?

File: Ethereum/Voting/_Quadratic_Voting.sol [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    for (uint i = 0; i < proposalNames.length; i++){
+        proposals.push(Proposal({
+        name: proposalNames[i],
+        voteCount: 0, voteCreditCount: 0
+        }));
+    }
+}
+
+
```
Comment (vishakh) [inline Ethereum/Voting/_Quadratic_Voting.sol:50]: Could you make the spacing between functions consistent?

**Review by vishakh — CHANGES_REQUESTED** [review]

File: Ethereum/Voting/_Quadratic_Voting.sol (hunk 3 above)
Comment (dwariyar) [reply Ethereum/Voting/_Quadratic_Voting.sol]: removed it

File: Ethereum/Voting/_Quadratic_Voting.sol (hunk 4 above)
Comment (dwariyar) [reply Ethereum/Voting/_Quadratic_Voting.sol]: changed spacing

**Review by vishakh — APPROVED** [review]

**Pull request merged**